src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +216,249 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
         } else {
             mh = target;
         }
-        ClassLoader proxyLoader = intfc.getClassLoader();
-        if (proxyLoader == null) {
-            ClassLoader cl = Thread.currentThread().getContextClassLoader(); // avoid use of BCP
-            proxyLoader = cl != null ? cl : ClassLoader.getSystemClassLoader();
-        }
-        final Method[] methods = getSingleNameMethods(intfc);
-        if (methods == null)
-            throw newIllegalArgumentException("not a single-method interface", intfc.getName());
-        final MethodHandle[] vaTargets = new MethodHandle[methods.length];
-        for (int i = 0; i < methods.length; i++) {
-            Method sm = methods[i];
-            MethodType smMT = MethodType.methodType(sm.getReturnType(), sm.getParameterTypes());
-            MethodHandle checkTarget = mh.asType(smMT);  // make throw WMT
-            checkTarget = checkTarget.asType(checkTarget.type().changeReturnType(Object.class));
-            vaTargets[i] = checkTarget.asSpreader(Object[].class, smMT.parameterCount());
+
+        /*
+         * For each interface I define a new hidden class and pass the original and
+         * caller-sensitive method handles to its constructor.
+         *
+         * The hidden classes defined for I is defined in a dynamic module M
+         * which has access to the types referenced by the members of I including
+         * the parameter types, return type and exception types.
+         */
+        Lookup lookup = getProxyClassLookup(intfc);  // throws IllegalArgumentException
+        Object proxy;
+        try {
+            MethodHandle constructor = lookup.findConstructor(lookup.lookupClass(), MT_void_Lookup_MethodHandle_MethodHandle)
+                    .asType(MT_Object_Lookup_MethodHandle_MethodHandle);
+            proxy = constructor.invokeExact(lookup, target, mh);
+        } catch (Throwable ex) {
+            throw uncaughtException(ex);
         }
-        final InvocationHandler ih = new InvocationHandler() {
-                private Object getArg(String name) {
-                    if ((Object)name == "getWrapperInstanceTarget")  return target;
-                    if ((Object)name == "getWrapperInstanceType")    return intfc;
-                    throw new AssertionError();
+        assert proxy.getClass().getModule().isNamed() : proxy.getClass() + " " + proxy.getClass().getModule();
+        return intfc.cast(proxy);
+    }
+
+    private record LocalMethodInfo(MethodTypeDesc desc, List<ClassDesc> thrown, String fieldName) {}
```

Comment:
Perhaps a simpler name `MethodInfo`

```suggestion
    private record MethodInfo(MethodTypeDesc desc, List<ClassDesc> thrown, String fieldName) {}
```